versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicImpl.java

```diff
@@ -0,0 +1,636 @@
+
+    if (!actAsAlreadyDeleted) {
+      persist.markReferenceAsDeleted(reference);
+    }
+
+    commitDeleteReference(reference);
+
+    persist.purgeReference(reference);
+
+    if (actAsAlreadyDeleted) {
+      // A previous deleteReference failed, act as if the first one succeeded, therefore this
+      // one would have not found the reference.
+      throw new RefNotFoundException(reference(name, expectedPointer, false));
+    }
+  }
+
+  static final class CommitReferenceResult {
+    final Reference reference;
+    final Kind kind;
+
+    CommitReferenceResult(Reference reference, Kind kind) {
+      this.reference = reference;
+      this.kind = kind;
+    }
+
+    enum Kind {
+      ADDED_TO_INDEX,
+      REF_ROW_EXISTS,
+      REF_ROW_MISSING
+    }
+  }
+
+  @VisibleForTesting // needed to simulate recovery scenarios
+  CommitReferenceResult commitCreateReference(String name, ObjId pointer)
+      throws RetryTimeoutException {
+    try {
+      return commitRetry(
+          persist,
+          (p, retryState) -> {
+            Reference refRefs = requireNonNull(p.fetchReference(REF_REFS.name()));
+            long created = p.config().currentTimeMicros();
+            RefObj ref = ref(name, pointer, created);
+            try {
+              p.storeObj(ref);
+            } catch (ObjTooLargeException e) {
+              throw new RuntimeException(e);
+            }
+
+            StoreKey k = key(name);
+
+            Instant now = persist.config().clock().instant();
+            CreateCommit c =
+                newCommitBuilder()
+                    .message("Create reference " + name + " pointing to " + pointer)
+                    .parentCommitId(refRefs.pointer())
+                    .addAdds(commitAdd(k, 0, requireNonNull(ref.id()), null, null))
+                    .headers(
+                        newCommitHeaders()
+                            .add("operation", "create")
+                            .add("name", name)
+                            .add("head", pointer.toString())
+                            .add("timestamp", now.toString())
+                            .add("timestamp.millis", Long.toString(now.toEpochMilli()))
+                            .build())
+                    .commitType(CommitType.INTERNAL)
+                    .build();
+
+            commitReferenceChange(p, refRefs, c);
+
+            return new CommitReferenceResult(reference(name, pointer, false), ADDED_TO_INDEX);
+          });
+    } catch (CommitConflictException e) {
+      checkState(e.conflicts().size() == 1, "Unexpected amount of conflicts %s", e.conflicts());
+
+      CommitConflict conflict = e.conflicts().get(0);
+      checkState(conflict.conflictType() == KEY_EXISTS, "Unexpected conflict type %s", conflict);
+
+      Supplier<StoreIndex<CommitOp>> indexSupplier = createRefsIndexSupplier();
+      StoreIndexElement<CommitOp> el = indexSupplier.get().get(key(name));
+      checkNotNull(el, "Key %s missing in index", name);
```

Comment:
> not be the same commit that caused the conflict

Yes. the intent here is to _handle_ the concurrent-create case, and otherwise throw.

What you describe is a create _and_ delete happening concurrent to "our" create. It'll be handled, but not gracefully.

Reply:
I mean that `el` may be `null` under normal concurrent modifications... Could we handle that gracefully, as opposed to throwing from `checkNotNull`?

Reply:
Not without a ton of code and tons of `if`s checking for all likely and unlikely race conditions.